File: backend/app/api/routes/batches.py
from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy import func, select
from sqlalchemy.orm import Session, selectinload

from app.core.database import get_db
from app.core.dependencies import get_current_user, require_roles
from app.models.batch import Batch, StudentBatch
from app.models.enums import UserRole
from app.models.student import Student
from app.models.user import User
from app.schemas.batch import BatchCreate, BatchResponse, BatchStudentResponse, BatchUpdate
from app.schemas.common import PaginatedResponse

router = APIRouter()


@router.post("/", response_model=BatchResponse, status_code=status.HTTP_201_CREATED)
def create_batch(
    payload: BatchCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_roles(UserRole.ADMIN, UserRole.TEACHER)),
) -> Batch:
    batch = Batch(
        institute_id=current_user.institute_id,
        name=payload.name,
        course=payload.course,
        schedule=payload.schedule,
        teacher_id=payload.teacher_id,
        start_date=payload.start_date,
        end_date=payload.end_date,
        fee_plan_id=payload.fee_plan_id,
    )
    db.add(batch)
    db.commit()
    db.refresh(batch)
    return batch


@router.get("/", response_model=PaginatedResponse)
def list_batches(
    search: str | None = Query(default=None),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=10, ge=1, le=200),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> PaginatedResponse:
    stmt = select(Batch).where(Batch.institute_id == current_user.institute_id)
    if current_user.role == UserRole.STUDENT:
        if not current_user.student_id:
            raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Student profile not linked")
        stmt = stmt.join(StudentBatch, StudentBatch.batch_id == Batch.id).where(
            StudentBatch.student_id == current_user.student_id
        )

    if search:
        stmt = stmt.where(Batch.name.ilike(f"%{search}%"))

    stmt = stmt.distinct()
    total = db.scalar(select(func.count()).select_from(stmt.subquery())) or 0
    rows = list(
        db.scalars(
            stmt.order_by(Batch.created_at.desc()).offset((page - 1) * page_size).limit(page_size)
        )
    )
    return PaginatedResponse(
        total=total,
        page=page,
        page_size=page_size,
        items=[BatchResponse.model_validate(item).model_dump(mode="json") for item in rows],
    )


@router.get("/{batch_id}", response_model=BatchResponse)
def get_batch(
    batch_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> Batch:
    stmt = select(Batch).where(Batch.id == batch_id, Batch.institute_id == current_user.institute_id)
    if current_user.role == UserRole.STUDENT:
        if not current_user.student_id:
            raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Student profile not linked")
        stmt = stmt.join(StudentBatch, StudentBatch.batch_id == Batch.id).where(
            StudentBatch.student_id == current_user.student_id
        )
    batch = db.scalar(stmt)
    if not batch:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Batch not found")
    return batch


@router.patch("/{batch_id}", response_model=BatchResponse)
def update_batch(
    batch_id: int,
    payload: BatchUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_roles(UserRole.ADMIN, UserRole.TEACHER)),
) -> Batch:
    batch = db.scalar(select(Batch).where(Batch.id == batch_id, Batch.institute_id == current_user.institute_id))
    if not batch:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Batch not found")
    updates = payload.model_dump(exclude_unset=True)
    for key, value in updates.items():
        setattr(batch, key, value)
    db.commit()
    db.refresh(batch)
    return batch


@router.delete("/{batch_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_batch(
    batch_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(require_roles(UserRole.ADMIN)),
) -> None:
    batch = db.scalar(select(Batch).where(Batch.id == batch_id, Batch.institute_id == current_user.institute_id))
    if not batch:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Batch not found")
    db.delete(batch)
    db.commit()
    return None


@router.get("/{batch_id}/students", response_model=list[BatchStudentResponse])
def list_batch_students(
    batch_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> list[BatchStudentResponse]:
    batch = db.scalar(select(Batch).where(Batch.id == batch_id, Batch.institute_id == current_user.institute_id))
    if not batch:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Batch not found")

    stmt = (
        select(Student)
        .join(StudentBatch, StudentBatch.student_id == Student.id)
        .where(StudentBatch.batch_id == batch_id, Student.institute_id == current_user.institute_id)
        .order_by(Student.full_name.asc())
    )
    students = list(db.scalars(stmt))
    return [
        BatchStudentResponse(id=s.id, full_name=s.full_name, phone=s.phone, email=s.email)
        for s in students
    ]


@router.get("/{batch_id}/schedule")
def batch_schedule(
    batch_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> dict:
    batch = db.scalar(
        select(Batch)
        .options(selectinload(Batch.teacher))
        .where(Batch.id == batch_id, Batch.institute_id == current_user.institute_id)
    )
    if not batch:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Batch not found")
    if current_user.role == UserRole.STUDENT:
        link = db.scalar(
            select(StudentBatch).where(
                StudentBatch.batch_id == batch.id,
                StudentBatch.student_id == current_user.student_id,
            )
        )
        if not link:
            raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Not allowed")
    return {
        "batch_id": batch.id,
        "name": batch.name,
        "course": batch.course,
        "schedule": batch.schedule,
        "teacher_name": batch.teacher.full_name if batch.teacher else None,
        "start_date": batch.start_date,
        "end_date": batch.end_date,
    }

File: backend/src/routes/batches.js
const express = require("express");
const { connectDb } = require("../db");
const { requireAuth, requireRoles } = require("../auth");
const { parsePagination } = require("./helpers");

const router = express.Router();

router.post("/batches", requireRoles("ADMIN", "TEACHER"), async (req, res) => {
  const db = await connectDb();
  const result = await db.run(
    `INSERT INTO batches (institute_id, name, course, schedule, teacher_id, start_date, end_date, fee_plan_id)
     VALUES (?, ?, ?, ?, ?, ?, ?, ?)`,
    [
      req.user.institute_id,
      req.body.name,
      req.body.course,
      req.body.schedule,
      req.body.teacher_id || null,
      req.body.start_date,
      req.body.end_date || null,
      req.body.fee_plan_id || null
    ]
  );
  const row = await db.get("SELECT * FROM batches WHERE id = ?", [result.lastID]);
  res.status(201).json(row);
});

router.get("/batches", requireAuth, async (req, res) => {
  const db = await connectDb();
  const { page, pageSize, offset } = parsePagination(req.query, 10);
  const params = [req.user.institute_id];
  let joinSql = "";
  let where = "WHERE b.institute_id = ?";
  if (req.user.role === "STUDENT") {
    if (!req.user.student_id) {
      return res.status(403).json({ detail: "Student profile not linked" });
    }
    joinSql = "JOIN student_batches sb ON sb.batch_id = b.id";
    where += " AND sb.student_id = ?";
    params.push(req.user.student_id);
  }
  if (req.query.search) {
    where += " AND b.name LIKE ?";
    params.push(`%${req.query.search}%`);
  }
  const total = await db.get(
    `SELECT COUNT(*) as total FROM (SELECT DISTINCT b.id FROM batches b ${joinSql} ${where}) x`,
    params
  );
  const rows = await db.all(
    `SELECT DISTINCT b.* FROM batches b ${joinSql} ${where}
     ORDER BY b.created_at DESC LIMIT ? OFFSET ?`,
    [...params, pageSize, offset]
  );
  res.json({ total: total.total, page, page_size: pageSize, items: rows });
});

router.get("/batches/:batchId", requireAuth, async (req, res) => {
  const db = await connectDb();
  const batchId = Number(req.params.batchId);
  let row;
  if (req.user.role === "STUDENT") {
    if (!req.user.student_id) {
      return res.status(403).json({ detail: "Student profile not linked" });
    }
    row = await db.get(
      `SELECT b.* FROM batches b
       JOIN student_batches sb ON sb.batch_id = b.id
       WHERE b.id = ? AND b.institute_id = ? AND sb.student_id = ?`,
      [batchId, req.user.institute_id, req.user.student_id]
    );
  } else {
    row = await db.get("SELECT * FROM batches WHERE id = ? AND institute_id = ?", [batchId, req.user.institute_id]);
  }
  if (!row) {
    return res.status(404).json({ detail: "Batch not found" });
  }
  res.json(row);
});

router.patch("/batches/:batchId", requireRoles("ADMIN", "TEACHER"), async (req, res) => {
  const db = await connectDb();
  const batchId = Number(req.params.batchId);
  const exists = await db.get("SELECT id FROM batches WHERE id = ? AND institute_id = ?", [batchId, req.user.institute_id]);
  if (!exists) {
    return res.status(404).json({ detail: "Batch not found" });
  }
  const updates = { ...req.body };
  Object.keys(updates).forEach((key) => updates[key] === undefined && delete updates[key]);
  if (Object.keys(updates).length) {
    const setSql = Object.keys(updates).map((key) => `${key} = ?`).join(", ");
    await db.run(`UPDATE batches SET ${setSql}, updated_at = CURRENT_TIMESTAMP WHERE id = ?`, [
      ...Object.values(updates),
      batchId
    ]);
  }
  const updated = await db.get("SELECT * FROM batches WHERE id = ?", [batchId]);
  res.json(updated);
});

router.delete("/batches/:batchId", requireRoles("ADMIN"), async (req, res) => {
  const db = await connectDb();
  const batchId = Number(req.params.batchId);
  const exists = await db.get("SELECT id FROM batches WHERE id = ? AND institute_id = ?", [batchId, req.user.institute_id]);
  if (!exists) {
    return res.status(404).json({ detail: "Batch not found" });
  }
  await db.run("DELETE FROM batches WHERE id = ?", [batchId]);
  res.status(204).send();
});

router.get("/batches/:batchId/students", requireAuth, async (req, res) => {
  const db = await connectDb();
  const batchId = Number(req.params.batchId);
  const batch = await db.get("SELECT id FROM batches WHERE id = ? AND institute_id = ?", [batchId, req.user.institute_id]);
  if (!batch) {
    return res.status(404).json({ detail: "Batch not found" });
  }
  const rows = await db.all(
    `SELECT s.id, s.full_name, s.phone, s.email
     FROM students s JOIN student_batches sb ON sb.student_id = s.id
     WHERE sb.batch_id = ? AND s.institute_id = ? ORDER BY s.full_name ASC`,
    [batchId, req.user.institute_id]
  );
  res.json(rows);
});

router.get("/batches/:batchId/schedule", requireAuth, async (req, res) => {
  const db = await connectDb();
  const batchId = Number(req.params.batchId);
  const batch = await db.get(
    `SELECT b.*, u.full_name AS teacher_name
     FROM batches b LEFT JOIN users u ON u.id = b.teacher_id
     WHERE b.id = ? AND b.institute_id = ?`,
    [batchId, req.user.institute_id]
  );
  if (!batch) {
    return res.status(404).json({ detail: "Batch not found" });
  }
  if (req.user.role === "STUDENT") {
    const link = await db.get("SELECT id FROM student_batches WHERE batch_id = ? AND student_id = ?", [
      batchId,
      req.user.student_id
    ]);
    if (!link) {
      return res.status(403).json({ detail: "Not allowed" });
    }
  }
  res.json({
    batch_id: batch.id,
    name: batch.name,
    course: batch.course,
    schedule: batch.schedule,
    teacher_name: batch.teacher_name,
    start_date: batch.start_date,
    end_date: batch.end_date
  });
});

module.exports = router;
